--- src/event/ngx_event_openssl.c
+++ src/event/ngx_event_openssl.c
@@ -63,4 +63,7 @@
      u_char *id, int len, int *copy);
  static void ngx_ssl_remove_session(SSL_CTX *ssl, ngx_ssl_session_t *sess);
+#if (OPENSSL_VERSION_NUMBER >= 0x10101000L)
+static void ngx_ssl_keylog_callback(const SSL *ssl, const char *line);
+#endif
  static void ngx_ssl_expire_sessions(ngx_ssl_session_cache_t *cache,
      ngx_slab_pool_t *shpool, ngx_uint_t n);
 
@@ -335,4 +338,36 @@
 
+#if (OPENSSL_VERSION_NUMBER >= 0x10101000L)
+
+static void
+ngx_ssl_keylog_callback(const SSL *ssl, const char *line)
+{
+    char              *path;
+    FILE              *fp;
+    ngx_connection_t  *c;
+
+    path = getenv("SSLKEYLOGFILE");
+
+    if (path == NULL) {
+        return;
+    }
+
+    fp = fopen(path, "a");
+
+    if (fp == NULL) {
+        c = ngx_ssl_get_connection((ngx_ssl_conn_t *) ssl);
+        if (c && c->log) {
+            ngx_log_error(NGX_LOG_CRIT, c->log, ngx_errno,
+                          "Could not open keylog file: \"%s\"", path);
+        }
+        return;
+    }
+
+    fprintf(fp, "%s\n", line);
+
+    fclose(fp);
+}
+
+#endif
+
+
 ngx_int_t
 ngx_ssl_create(ngx_ssl_t *ssl, ngx_uint_t protocols, void *data)
 {
@@ -342,6 +377,12 @@
         return NGX_ERROR;
     }
 
+#if (OPENSSL_VERSION_NUMBER >= 0x10101000L)
+    if (getenv("SSLKEYLOGFILE")) {
+        SSL_CTX_set_keylog_callback(ssl->ctx, ngx_ssl_keylog_callback);
+    }
+#endif
+
     if (SSL_CTX_set_ex_data(ssl->ctx, ngx_ssl_server_conf_index, data) == 0) {
         ngx_ssl_error(NGX_LOG_EMERG, ssl->log, 0,
